src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/oops/Annotations.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+package sun.jvm.hotspot.oops;
+
+import java.io.*;
+import java.util.*;
+import sun.jvm.hotspot.debugger.*;
+import sun.jvm.hotspot.interpreter.Bytecodes;
+import sun.jvm.hotspot.runtime.*;
+import sun.jvm.hotspot.types.*;
+import sun.jvm.hotspot.utilities.*;
+import sun.jvm.hotspot.utilities.Observable;
+import sun.jvm.hotspot.utilities.Observer;
+
+// An Annotation is an oop containing annotations as described in the class file
+
+public class Annotations extends Metadata {
+  static {
+    VM.registerVMInitializedObserver(new Observer() {
+        public void update(Observable o, Object data) {
+          initialize(VM.getVM().getTypeDataBase());
+        }
+      });
+  }
+
+  public Annotations(Address addr) {
+    super(addr);
+  }
+
+  public boolean isAnnotations()            { return true; }
+
+  private static synchronized void initialize(TypeDataBase db) throws WrongTypeException {
+    Type type = db.lookupType("Annotations");
+    classAnnotations = type.getAddressField("_class_annotations");
+    fieldsAnnotations = type.getAddressField("_fields_annotations");
+    classTypeAnnotations = type.getAddressField("_class_type_annotations");
+    fieldsTypeAnnotations = type.getAddressField("_fields_type_annotations");
+  }
+
+  private static AddressField classAnnotations;
+  private static AddressField fieldsAnnotations;
+  private static AddressField classTypeAnnotations;
+  private static AddressField fieldsTypeAnnotations;
+
+  public U1Array getClassAnnotations() {
+    Address addr = classAnnotations.getValue(getAddress());
+    return VMObjectFactory.newObject(U1Array.class, addr);
+  }
+
+  public U1Array getFieldAnnotations(int fieldIndex) {
+    Address addr = fieldsAnnotations.getValue(getAddress());
+    ArrayOfU1Array annotationsArray = VMObjectFactory.newObject(ArrayOfU1Array.class, addr);
```

Comment:
How about caching this result so you don't need to allocate a new object every time this API is called. Same thing in `getFieldTypeAnnotations()`.

Reply:
I think VMObjectFactory is a better place to implement the caching behavior so that all such patterns can benefit from it. I think it is better addressed in another task.

Reply:
I think maybe you misunderstood what I meant by "cache". I'm not talking about a hashmap of weak references or anything like that. Just add a `ArrayOfU1Array annotationsArray` field to the Annotations object and store the result there.

Reply:
Got it. Updated the code as suggested.